checker/src/options.rs

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+//
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+use clap::{App, AppSettings, Arg};
+use itertools::Itertools;
+use shellwords;
+
+/// Creates the clap::App metadata for argument parsing.
+fn make_options_parser<'a>() -> App<'a, 'a> {
+    // We could put this into lazy_static! with a Mutex around, but we really do not expect
+    // to construct this more then once per regular program run.
+    App::new("MIRAI")
+    .setting(AppSettings::NoBinaryName)
+    .version("v1.0.5")
+    .arg(Arg::with_name("single_func")
+        .long("single_func")
+        .takes_value(true)
+        .help("Focus analysis on the named function.")
+        .long_help("Name is the simple name of a top-level crate function or a MIRAI summary key."))
+    .arg(Arg::with_name("test_only")
+        .long("test_only")
+        .takes_value(false)
+        .help("Focus analysis on #[test] methods.")
+        .long_help("Only #[test] methods and their usage are analyzed. This must be used together with the rustc --test option."))
+    .arg(Arg::with_name("flaky")
+        .long("flaky")
+        .takes_value(false)
+        .help("Report diagnostics even with incomplete information.")
+        .long_help("This removes measurements to suppress flaky diagnostics. The effect is implementation dependent."))
+}
+
+/// Represents options passed to MIRAI.
+#[derive(Debug, Default)]
+pub struct Options {
+    pub single_func: Option<String>,
+    pub test_only: bool,
+    pub flaky: bool,
+}
+
+impl Options {
+    /// Parse options from an argument string. The argument string will be split using unix
+    /// shell escaping rules. Any content beyond the leftmost `--` token will be returned
+    /// (excluding this token).
+    pub fn parse_from_str(&mut self, s: &str) -> Vec<String> {
+        self.parse(&shellwords::split(s).unwrap())
```

Comment:
Only use unwrap when MIRAI can prove it will never fail. If this is an assumption, use expect.

In this case I suspect things can fail because of bad user input. In such cases I want to see a proper user error message.

Reply:
I wish the code would have systematic error handling and user messaging setup. Instead, we are using unstable rustc APIs for which documentation is hard to find, and writing into logs which should arguably user messages. There should be a diag module somewhere where all the needed functionality is tugged away in an easy discoverable way, then it would be easier to follow rules.

In any case, I'm going to call early_error here now.